design-decisions/2020-05-02-function-rules.md

```diff
@@ -0,0 +1,176 @@
+# Interpretation of function rules
+
+## Purpose
+
+The purpose of this document is to describe in detail how K function rules will be interpreted in Kore.
+We will also summarize the motivation for this decision, that is:
+what problems this decision solves.
+
+## Notation
+
+`Xᵢ` and `Yᵢ` refer to element variables.
+`{X}` and `{Y}` refer to all variables in the respective family of element variables.
+`fun(...)` is a function symbol.
+`φ` are arbitrary patterns appearing as the arguments of a function rule.
+`ψ` are arbitrary patterns appearing on the right-hand side of function rules.
+`Pre` and `Post` are respectively pre- and post-condition patterns.
+
+## Background
+
+The family of K rules defining `fun`,
+
+```.k
+rule fun(φ₁ᵢ({Y}), φ₂ᵢ({Y}), ...) => ψᵢ({Y}) requires Preᵢ({Y}) ensures Postᵢ({Y})
+```
+
+is interpreted in Kore as
+
+```.kore
+axiom \implies(Preᵢ({Y}), (fun(φ₁ᵢ({Y}), φ₂ᵢ({Y}), ...) = ψᵢ({Y})) ∧ Postᵢ({Y})).
+```
+
+## Problems
+
+### Partial heads
+
+See also: [kframework/kore#1146](https://github.com/kframework/kore/issues/1146)
+
+Rules with partial heads are potentially unsound.
+Consider the family of rules
+
+```.k
+rule sizeMap(_ |-> _ M:Map) = 1 +Int sizeMap(M)
+rule sizeMap(_ |-> _) = 1
+rule sizeMap(.Map) = 0
+```
+
+producing axioms
+
+```.kore
+axiom sizeMap(concatMap(elementMap(_1, _2), M:Map)) = +Int(1, sizeMap(M))
+axiom sizeMap(elementMap(_, _)) = 1
+axiom sizeMap(.Map) = 0.
+```
+
+Note that the first rule has a partial head;
+we can instantiate it at `M = elementMap(_1, _2)` to prove
+
+```.kore
+sizeMap(concatMap(elementMap(_1, _2), elementMap(_1, _2))) = +Int(1, sizeMap(M))
+sizeMap(\bottom) = +Int(1, sizeMap(elementMap(_1, _2)))
+sizeMap(\bottom) = +Int(1, 1)
+\bottom = 2.
+```
+
+### Or patterns
+
+See also: [kframework/kore#1245](https://github.com/kframework/kore/issues/1245)
+
+The user may write an or-pattern rule such as
+
+```.k
+rule fun(A #Or B) => C
+```
+
+as shorthand for two rules,
+
+```.k
+rule fun(A) = C
+rule fun(B) = C
+```
+
+Under the current interpretation, this produces an axiom
+
```

Comment:
Why?  The current interpretation does not talk about merging rules (btw, `=` should be `=>` in the rules above).